File: HisabCore/Sources/HisabCore/Coverage.swift
import Foundation

/// Lightweight document projection the app layer feeds into grid derivation.
public struct DocumentSummary: Sendable {
    public let id: UUID
    public let source: Source
    public let period: DatePeriod

    public init(id: UUID, source: Source, period: DatePeriod) {
        self.id = id
        self.source = source
        self.period = period
    }
}

public enum CellState: Equatable, Sendable {
    case present(documentIDs: [UUID])
    case awaiting
}

/// The month × source presence grid. Months are derived, never stored: the union of every
/// document's covered months and manual pins, gap-filled so interior holes stay visible.
public struct CoverageGrid: Sendable {
    public let months: [YearMonth]  // newest first
    private let cells: [YearMonth: [Source: [UUID]]]

    public func state(month: YearMonth, source: Source) -> CellState {
        if let ids = cells[month]?[source], !ids.isEmpty {
            return .present(documentIDs: ids)
        }
        return .awaiting
    }

    public static func derive(documents: [DocumentSummary], pinnedMonths: Set<YearMonth>) -> CoverageGrid {
        var cells: [YearMonth: [Source: [UUID]]] = [:]
        var touched = pinnedMonths
        for doc in documents {
            for month in doc.period.months {
                touched.insert(month)
                cells[month, default: [:]][doc.source, default: []].append(doc.id)
            }
        }
        guard let lo = touched.min(), let hi = touched.max() else {
            return CoverageGrid(months: [], cells: [:])
        }
        let months = YearMonth.months(from: lo, through: hi).reversed()
        return CoverageGrid(months: Array(months), cells: cells)
    }
}

File: hisab_flutter/packages/hisab_core/lib/src/coverage.dart
/// Month × source coverage grid. Port of Coverage.swift.
library;

import 'domain.dart';
import 'year_month.dart';

class DocumentSummary {
  final String id;
  final Source source;
  final DatePeriod period;
  const DocumentSummary(
      {required this.id, required this.source, required this.period});
}

sealed class CellState {
  const CellState();
}

class CellPresent extends CellState {
  final List<String> documentIDs;
  const CellPresent(this.documentIDs);
}

class CellAwaiting extends CellState {
  const CellAwaiting();
}

/// Months are derived, never stored: the union of every document's covered
/// months and manual pins, gap-filled so interior holes stay visible.
class CoverageGrid {
  final List<YearMonth> months; // newest first
  /// Sources actually observed in the input documents: payment apps first,
  /// then banks, alphabetical within kind.
  final List<Source> sources;
  final Map<YearMonth, Map<Source, List<String>>> _cells;

  CoverageGrid._(this.months, this.sources, this._cells);

  CellState state(YearMonth month, Source source) {
    final ids = _cells[month]?[source];
    if (ids != null && ids.isNotEmpty) return CellPresent(ids);
    return const CellAwaiting();
  }

  static CoverageGrid derive({
    required List<DocumentSummary> documents,
    required Set<YearMonth> pinnedMonths,
  }) {
    final cells = <YearMonth, Map<Source, List<String>>>{};
    final touched = Set<YearMonth>.from(pinnedMonths);
    for (final doc in documents) {
      for (final month in doc.period.months) {
        touched.add(month);
        cells.putIfAbsent(month, () => {})
            .putIfAbsent(doc.source, () => [])
            .add(doc.id);
      }
    }
    final sources = Source.ordered(documents.map((d) => d.source));
    if (touched.isEmpty) return CoverageGrid._([], sources, {});
    final sorted = touched.toList()..sort();
    final months = YearMonth.monthsFromThrough(sorted.first, sorted.last)
        .reversed
        .toList();
    return CoverageGrid._(months, sources, cells);
  }
}
